src/preact/base-element.js

```diff
@@ -610,6 +629,107 @@ export class PreactBaseElement extends AMP.BaseElement {
     }
     return this.defaultProps_[prop];
   }
+
+  /**
+   * Returns reference to upgraded imperative API object, as in React's
+   * useImperativeHandle.
+   * @return {!Promise<!API_TYPE>}
+   */
+  getApi() {
+    const api = this.apiWrapper_;
+    if (api) {
+      return Promise.resolve(api);
+    }
+    if (!this.deferredApi_) {
+      this.deferredApi_ = new Deferred();
+    }
+    return this.deferredApi_.promise;
+  }
+
+  /**
+   * Creates a wrapper around a Preact ref. The API surface exposed by this ref
+   * **should** be consistent accross all rerenders.
+   *
+   * This wrapper is necessary because every time React rerenders, it creates
+   * (depending on deps checking) a new imperative handle and sets that to
+   * `ref.current`. So if we ever returned `ref.current` directly, it could go
+   * stale by the time its actually used.
+   *
+   * @param {!API_TYPE} current
+   */
+  initApiWrapper_(current) {
+    const api = map();
+    const keys = Object.keys(current);
+    const ref = this.ref_;
+    for (let i = 0; i < keys.length; i++) {
+      const key = keys[i];
+      wrapRefProperty(ref, api, key);
+    }
+    this.apiWrapper_ = api;
+    if (this.deferredApi_) {
+      this.deferredApi_.resolve(api);
+      this.deferredApi_ = null;
+    }
+  }
+
+  /**
+   * Verifies that every Preact render exposes the same API surface as the previous render.
+   * If it does not, the API wrapper is syncrhonized.
+   *
+   * @param {!API_TYPE} current
+   */
+  checkApiWrapper_(current) {
+    const api = this.apiWrapper_;
+    const newKeys = Object.keys(current);
+    for (let i = 0; i < newKeys.length; i++) {
+      const key = newKeys[i];
+      if (!hasOwn(api, key)) {
+        wrapRefProperty(this.ref_, api, key);
+      }
+    }
+    const oldKeys = Object.keys(api);
+    for (let i = 0; i < oldKeys.length; i++) {
+      const key = oldKeys[i];
+      if (!hasOwn(current, key)) {
+        delete api[key];
+      }
+    }
+  }
+}
+
+/**
+ * @param {!{current: *}} ref
+ * @param {!Object} api
+ * @param {string} key
+ */
+function wrapRefProperty(ref, api, key) {
+  Object.defineProperty(api, key, {
+    configurable: true,
+
+    get() {
+      return ref.current[key];
+    },
+
+    set(v) {
```

Comment:
When would we ever call `set`?

Reply:
Some API's could, eg [`video.currentTime = 100`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/currentTime)

Reply:
and we would want to directly expose the `video` object via `getApi()`?

Reply:
I don't understand. Wouldn't this `set` assign to the ref itself as opposed to a member of the ref. For example even if we directly exposed `video` from the AmpVideo component via `getAPI().video`, we still wouldn't need the setter.

Reply:
`video` isn't exposed as a prop here, `video` _is_ the API object. In a real `<video>` element, you do `el.currentTime = 100`. In Bento, you'd do `api.currentTime = 100`. Eg,

```js
useImperativeHandle(ref, () => {
  return {
    get currentTime() {
      return time;
    },
    set currentTime(t) {
      time = t; // update video element here…
    }
  };
});

// This invokes a setter.
ref.currentTime = 100;

Reply:
I see. we just don't expose set yet, but likely will/should.

https://github.com/ampproject/amphtml/blob/4209fe94e53f4a4d2b345ef39056b0e4525c5e1d/extensions/amp-video/1.0/video-wrapper.js#L158-L163

Reply:
Thank you for the explanation!